app/models/trigger_metadata_event_callback.rb

```diff
@@ -1,8 +1,11 @@
 class TriggerMetadataEventCallback
   def after_save(model)
+    env = environment(model)
```

Comment:
So there something strange here, it seems this logic is not needed b'cos 
```
def environment(model)
    if model.aggregate.respond_to?(:component)
      model.aggregate.component.environment
    elsif model.respond_to?(:service)
      if model.service.sp_component_id.present?
        model.service.sp_component.environment
      elsif model.service.msa_component_id.present?
        model.service.msa_component.environment
      end
    else
      model.aggregate.environment
    end
  end
```
seems to have fixed the **nil** issue; but I don't understand how? Apart from the nesting and .present? it looks equivalent to the original logic

Reply:
Try 
```
PublishServicesMetadataEvent.create(	    PublishServicesMetadataEvent.create(
      event_id: model.id,	      event_id: model.id,
      environment: environment(model),	 
    )
```
And see if you test pass

Reply:
I'd say don't mess with this ... it works and it took a long time to get nail the logic so that it works.  The specific change is to look at if the id is actually present as services always have a sp_component_id and msa_componenent_id even if they are nil.

Reply:
I think I know how it works it b'cos your model can either be from the `if model.respond_to?(:service)`
```
AssignMsaComponentToServiceEvent
assign_sp_component_to_service_event
and now
ChangeServiceEvent
```
before there was
```
model.respond_to?(:sp_component_id)
model.respond_to?(:msa_component_id)
```
which excluded the **ChangeServiceEvent** that was where the nil was generated